### PR TITLE
Plot scores for combined predictions

### PR DESCRIPTION
**Pull request by LTLA**
Making this into a PR to keep track of the edits.

Firstly, move stuff in `utils-Viz.R` into `utils.R`, that's why it's there. I usually only split it up into smaller `utils-*.R` files if there are too many utilities, and that's not (yet) the case here.

Secondly, I would prefer if the plotting logic were isolated from the data-extraction logic. You can see how the various `*.use` arguments are sprinkled throughout the body of the function, which makes it difficult to keep track of its effects. Instead, I would suggest the following:

- define an internal helper function, say, `.plot_score_distribution`, for which the only job is to make plots. It takes the `scores.values=`, `calls.values=` and `prune.values=` arguments (possibly better names, but whatever) that take the actual matrix/vector of values to be used. You could probably get rid of the need to take the `results=` argument directly in this helper.
- Have `plotScoreDistribution` be responsible for extracting values from `results=` and providing it into the internal helper. You can see how this isolates the effects of extraction from the effects of plotting, making it easy to keep track of which function is doing what and when.

Same arguments apply for `plotScoreHeatmap`.

### CODE REVIEW

**Comment by dtm2451** [comment]
- [x] Merge utils-Viz.R in utils.R
- Reorganize code to separate data extraction and plotting logic:
  - [x] `plotScoreDistribution`
  - [ ] `plotScoreHeatmap`

**Comment by dtm2451** [comment]
Minor note: In making the `.plot_score_distribution` sub-function, I made a change to the plotting order for the purpose of code readability. Outcome is that now cutoff lines are drawn in front regardless of `'dots.on.top' = TRUE/FALSE`. I think it's fine as the purpose of that argument is more for dots versus violin relation -- it allows for dots to be put behind the violins when they drown out the violins -- but chime in if you disagree. We could change the name of that input `violin.in.back` 🤷but I don't think this is worth any potential code breaks.

File: R/plotScoreHeatmap.R [hunk 1]
```diff
@@ -199,9 +259,45 @@ plotScoreHeatmap <- function(results, cells.use = NULL, labels.use = NULL,
         args$color <- color
     }
 
+    # Troubleshooting and testing purposes
+    if (!is.null(args$return.data) && args$return.data) {
+        return(args)
+    }
```
Comment (dtm2451) [inline R/plotScoreHeatmap.R:265]: I wonder your thoughts on this bit @LTLA. It'd give me a way to write unit tests for many of the things that currently require manually checking the plot outputs. It's also hidden from users within a sub-function & requires (undocumented) provision of `return.data = TRUE` within the initial `...`.

Comment (LTLA) [reply R/plotScoreHeatmap.R:265]: Yes, that's fine. Though don't go overboard with the unit tests, remember you'll be the one needing to change them all then the function changes.

Comment (dtm2451) [reply R/plotScoreHeatmap.R:265]: These tests are largely already written.  Just with this I can swap the 'expect_pheatmap' + manual check for a direct 'is_equal(output_piece, expected)' or the like.

**Comment by LTLA** [comment]
I was hoping for all of:

https://github.com/LTLA/SingleR/blob/5cb1823aab16a7d21d2e1667b04ea9354afcfd87/R/plotScoreHeatmap.R#L161-L234

To move into `.plot_score_heatmap`. This would mean that the only job - literally the only job! - of the outer `plotScoreHeatmap` function would be to switch between single- and multi-ref mode. 

This would consolidate the result-grabbing logic in `grab_results`, etc. as you wouldn't have a whole bunch of `is.null(results$orig.results)` calls scattered around the file. Importantly, it would separate the result-grabbing logic from the plotting logic, as previously discussed. For example, once I start reading `plot_score_heatmap`, I would know that I have everything necessary to create the plot, and I no longer have to care about where they came from in `results`.

If you are ambitious, you could write a single function that implements the result-grabbing logic and calls an arbitrary plotting function, allowing you to use the same result-grabbing logic across `plotScoreDistribution` and `plotScoreHeatmap`. This would be a function that accepts a plotting function as an argument and calls it; for example, see:

https://github.com/MarioniLab/scran/blob/575839dadb4a92276e79ddf1bda4ca9660fa544f/R/utils_markers.R#L54-L106

(This takes in an arbitrary DE testing function and applies a consistent set of transformations to its output, allowing me to re-use the same code for a variety of different marker detection methods.)

**Comment by dtm2451** [comment]
Hmmm.  I can certainly add even more isolation.  The result-grabbing parts are altogether at the top of the code you highlight.  (There's one other `is.null(results$orig.results)` lower down, but it's just for setting the plot title.  It can be grouped with the bits at the top instead.)  Everything from `annotation_col <- .make_annotation_col(` (line 175) to `### Build the heatmap` (line 226) are data manipulations.  But your comment makes it obvious that this delineation is not clear, so I will pull the rest of what's in your code grab into `.plot_score_heatmap` to make the responsibilities split more apparent.

I don't think it would be useful to make a results-grabbing wrapper for both `Distribution` and `Heatmap` as they can deal with drastically different results.  Biggest culprit of this is the `show` argument of `Distribution`, but there are some other differences too.  I'll circle back to this after I regroup `plotScoreHeatmap`.

**Comment by LTLA** [comment]
The general philosophy is, where we can, avoid doing repeated checks for `!is.null(results$orig.results)`. Ideally this would be decided once.

**Comment by dtm2451** [comment]
For computation, that's often possible.  I've found that it becomes harder, sometime impossible, when plot label creation or messages should also know such information.  Those bits are now tucked away within utils functions as much as possible.  At least for `plotScoreHeatmap`.

Ultimately, I shrunk the data manipulation bits of your quoted text into their own sub-functions.  I'm back to liking them better within the main function and not the more-*plot* specific `.plot_score_heatmap` subfunction.  👍 if you approve.

Turning back to `plotScoreDistribution` for now to see what else might be simplified there.

**Comment by LTLA** [comment]
See #111 into this branch for my vision of how this isolation would look.

**Comment by dtm2451** [comment]
#111 was great.  I've now reorganized `plotScoreDistribution` in a similar way.

**Comment by LTLA** [comment]
Looking good. Just a few points:

This seems extraneous:

https://github.com/LTLA/SingleR/blob/b0a930ee485a72fab0f88f99644a8979817e0641/R/plotScoreDistribution.R#L222-L230

`if` statements also have return values, so instead of:

https://github.com/LTLA/SingleR/blob/b0a930ee485a72fab0f88f99644a8979817e0641/R/plotScoreDistribution.R#L391-L393

... which has an ugly coercion to string, you might use:

```
which_pruned <- if (chosen.pruned==0) {
    ifelse(is.combined, "(in Final)", "")
} else {
    paste0("(in Ref #", chosen.pruned, ")")
}
```

One also consider showing the actual names of the `orig.results` instead of the indices. Right now these have silly default names if the `results` were not named in the `combine*Results()` call, but this is easy to change. Then the plot would respect the naming scheme, e.g., `Scores (HPCA)` or `Scores (Blueprint)`, which is a lot more useful than `Scores (ref1)`.

**Comment by LTLA** [comment]
Switched to showing the names of the supplied references. The default names are kind of boring but you can change the names of `results$orig.results` (either directly, or by changing the names of `results` passed to `combine*Results()`) to get some sweet sweet goodness.

blah

Also, mop up the `show.all.originals=` lying around in the examples.

**Comment by dtm2451** [comment]
I'd wanted the ref-names shown, just hadn't noticed that you'd started naming the `$orig.results` slots.  Last I'd checked, the names looked like a failed deparse.  I assume `SingleR()`'s ref list inputs can now be named?

**Comment by dtm2451** [comment]
In addition to the above suggestions, I:
- Shifted `.add_cutoff_lines()` altogether (the non-dup one) into `.plot_score_distribution`
- ~Update `.prune_label()`~ utilized `.values_title()` instead.
- Brought the examples in both `plotScores` functions up to date.

Running check now before committing.

**Comment by dtm2451** [comment]
For writing unit tests, should I copy a version of these lines into my `test-score-dist.R` and `test-heatmap.R` or should a version of this be made into a separate `setup-multiRef.R` file for use across tests?
https://github.com/LTLA/SingleR/blob/6a62473c4cf7b6a393c6d50ff3342fae0d29105e/tests/testthat/test-recomputed.R#L4-L26 

The "version" I've compiled for my current unit test plans is this:
```
### Setup a multi-ref output ###

ref <- .mockRefData(nreps=8)
ref1 <- ref[,1:4%%4==0]
ref1 <- ref1[,sample(ncol(ref1))]
ref1 <- scater::logNormCounts(ref1)

ref2 <- ref[,1:4%%4!=0]
ref2 <- ref2[,sample(ncol(ref2))]
ref2 <- scater::logNormCounts(ref2)

ref2$label <- tolower(ref2$label)

combined <- SingleR(
    test, ref = list(smallRef = ref1, largeRef = ref2),
    labels = list(ref1$label, ref2$label))
```

**Comment by dtm2451** [comment]
To gather these together and to make sure all are approved of...

User-visible changes made to single-result outputs:
- plotScoreDistribution:
  - `show = "scores"` by default instead of "delta.med"
  - `show.nmad = 3` by default instead of `NULL` to reflect default pruning parameter.
  - input name change: `labels` -> `labels.use`.  = more descriptive & consistent with `plotScoreHeatmap`
- plotScoreHeatmap:
  - color change for non-normalized plots (color blindness friendly)
  - titled as "Scores" (ref-name appended too for multi-ref)

**Comment by dtm2451** [comment]
Remaining tasks:
- [ ] update docs
- [ ] add the multi-ref setup chunk to `setup.R`
- [ ] update `NEWS`

**Comment by dtm2451** [comment]
Ready-to-go from my perspective.

**Comment by LTLA** [comment]
Do you have "squash and merge" ability? Best if you reorganized the commit message to whatever you thought was important.

**Pull request merged**